src/common/libsubprocess/local.c

```diff
@@ -178,6 +216,36 @@ static void local_stderr_cb (flux_reactor_t *r, flux_watcher_t *w,
     local_output (c, w, revents, c->p->ops.on_stderr);
 }
 
+static void initial_credits_cb (flux_reactor_t *r,
+                                flux_watcher_t *w,
+                                int revents,
+                                void *arg)
+{
+    flux_subprocess_t *p = arg;
+    struct subprocess_channel *c;
+
+    if (!p->ops.on_credit)
+        return;
+
+    c = zhash_first (p->channels);
+    while (c) {
+        if (c->flags & CHANNEL_WRITE) {
+            /* N.B. send cmd_option_bufsize() result not
+             * c->buffer_space, to avoid chance
+             * c->buffer_space already altered.
+             */
+            int bufsize = cmd_option_bufsize (p, c->name);
+            /* should have been checked at setup */
+            assert (bufsize > 0);
+            p->ops.on_credit (p, c->name, bufsize);
+        }
+        c = zhash_next (p->channels);
+    }
+
+    /* initial credits sent, we don't know need this watcher anymore */
+    flux_watcher_stop (p->initial_credits_w);
```

Comment:
Do we still need a dedicated watcher and code path for the initial callback or can fbuf callbacks drive the initial callback for each channel individually?

Reply:
I haven't figured out a good way to do it otherwise (doesn't mean there isn't a way, i just don't see it).  The key is that we have to return the subprocess object to the caller before we can send the initial credits.  So that suggests we need to setup something to send the initial credits after the subprocess object is returned to the caller.  A watcher is the obvious choice.

Reply:
My thought was have the fbuf write watcher do it, like have it call `ev_fbuf_write_wakeup()` on creation so as soon as the reactor runs, there is a callback?

Reply:
The `ev_fbuf_write_wakeup()` only wakes up the IO watcher.  And in order for the IO watcher callback to be triggered, there'd have to be activity on the file descriptor, which we don't control.

Reply:
Yeah but it's a POLLOUT and if we haven't written anything yet it's guaranteed(?) to fire.

Reply:
ahhh ok, I see where you're going with this now.

I'll have to ponder this and/or maybe give it a shot tomorrow.  This does feel a little hack-ish to me.  We're now creating an exception to start the IO watcher when there isn't data in the buffer to write out yet.

Also, per a prior discussion, I think we want the initial credits sent out via a prepare watcher.  The IO watcher callback is presumably a check watcher.

Reply:
Re: prepare watcher.  Well, early on I thought it was critical for the subprocess server to send `add-credit` responses right away to avoid contributing to the latency of remotely executing a fast running command.  Since then we discussed "borrowing credit" which would allow a small amount of data to be sent without waiting for the initial `add-credit` response.  So I think my position has changed on that.  Did you have a different reason for wanting to expedite the first callbacks?

Regarding hackishness, if the proper functioning of the fbuf write watcher were to depened on the `fd` passed into `fbuf_write_watcher_create()` being immediatley ready, that precondition could be checked with `poll(2)` in the create function and it could fail with EINVAL if it's not met.  Since it's a private function to libsubprocess and there is only one user, I don't think this will be a liability. In reality, if the fd is not ready, I think it would only delay the callback, which would still be "correct", so I don't think that check is really needed.l

If we do end up needing to use a different watcher for the first response, I suggest we put it down in the fbuf  write watcher not at the higher level.  Then at least it becomes part of the same abstraction that's activating the watcher when buffer space becomes available.  The first response is the same case really - all the buffer space is available.

Reply:
> Since then we discussed "borrowing credit" which would allow a small amount of data to be sent without waiting for the initial add-credit response. So I think my position has changed on that. Did you have a different reason for wanting to expedite the first callbacks?

I think the main reason is that a common use case will still be to start writing after the first credit response.  i.e. put all of the writing in the `on_credit()` callback.  So it feels like a good idea to send that quick.

> If we do end up needing to use a different watcher for the first response, I suggest we put it down in the fbuf write watcher not at the higher level. 

That seems like a good idea.

Reply:
Test cases aside, where performance isn't critical, what's the common case where the _data_ to write is always available and availability of buffer space is the sole driver?

Reply:
good point, I had been thinking of how I implemented things in #6370, but I had not tried using the "negative credits" idea in there yet.  So perhaps it's not as big a deal as I originally thought.

Reply:
Perhaps I should update the RFC PR to reflect some of the things we've discussed.  Progress!

Reply:
Update pushed to flux-framework/rfc#427.